modules/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaObserveSenderImpl.java

```diff
@@ -0,0 +1,701 @@
+                CaObserveSenderImpl.this.onApplyValChange(newVals.get(0));
+            }
+        });
+        
+        car = new CaCarRecord(carRecord, epicsService);
+        car.registerClidListener(carClidListener = (String arg0, List<Integer> newVals) -> {
+            if (newVals != null && !newVals.isEmpty()) {
+                CaObserveSenderImpl.this.onCarClidChange(newVals.get(0));
+            }
+        });
+        car.registerValListener(carValListener = (String arg0, List<CarState> newVals) -> {
+            if (newVals != null && !newVals.isEmpty()) {
+                CaObserveSenderImpl.this.onCarValChange(newVals.get(0));
+            }
+        });
+
+        if(stopCmd!=null && stopCmd.length()>0) {
+            stopMark = epicsReader.getShortChannel(stopCmd + CAD_MARK_SUFFIX);
+            stopMark.registerListener(stopMarkListener = (String arg0, List<Short> newVals) -> {
+                if (newVals != null && !newVals.isEmpty()) {
+                    CaObserveSenderImpl.this.onStopMarkChange(newVals.get(0));
+                }
+            });
+        } else {
+            stopMark = null;
+        }
+
+        if(abortCmd!=null && abortCmd.length()>0) {
+            abortMark = epicsReader.getShortChannel(abortCmd + CAD_MARK_SUFFIX);
+            abortMark.registerListener(abortMarkListener = (String arg0, List<Short> newVals) -> {
+                if (newVals != null && !newVals.isEmpty()) {
+                    CaObserveSenderImpl.this.onAbortMarkChange(newVals.get(0));
+                }
+            });
+        } else {
+            abortMark = null;
+        }
+
+        executor = new ScheduledThreadPoolExecutor(2);
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String getApply() {
+        return apply.getEpicsName();
+    }
+
+    @Override
+    public String getCAR() {
+        return car.getEpicsName();
+    }
+
+    void unbind() {
+        
+        executor.shutdown();
+
+        try {
+            apply.unregisterValListener(valListener);
+        } catch (CAException e) {
+            LOG.warning(e.getMessage());
+        }
+        try {
+            car.unregisterClidListener(carClidListener);
+        } catch (CAException e) {
+            LOG.warning(e.getMessage());
+        }
+        try {
+            car.unregisterValListener(carValListener);
+        } catch (CAException e) {
+            LOG.warning(e.getMessage());
+        }
+        if(stopMark!=null) {
+            try {
+                stopMark.unRegisterListener(stopMarkListener);
+            } catch (CAException e) {
+                LOG.warning(e.getMessage());
+            }
+        }
+        if(abortMark!=null) {
+            try {
+                abortMark.unRegisterListener(abortMarkListener);
+            } catch (CAException e) {
+                LOG.warning(e.getMessage());
+            }
+        }
+
+        apply.unbind();
```

Comment:
Can unbin throw?

Reply:
Well, it has a couple of asserts that should never trigger.